panic: Go driver while applying grpc dial options

We see this panic hundreds of times per day, and haven't been able to track down why. It has persisted since ~v5 until now on v15.0.2, and across a few dozen versions of the grpc package. We never see this happen anywhere else in our grpc stack, so I have no doubt that it is isolated to the Go vitessdriver usage.

cc @MordFustang21  

```go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x946854]

goroutine 147 [running]:
google.golang.org/grpc.(*funcDialOption).apply(0x18bc7f8?, 0xc000046038?)
	/Users/derek/go/pkg/mod/google.golang.org/grpc@v1.51.0/dialoptions.go:105 +0x14
google.golang.org/grpc.DialContext({0x18bc7f8, 0xc000046038}, {0xc001878bd0, 0x11}, {0xc006ce4000, 0x1d7e, 0x8?})
	/Users/derek/go/pkg/mod/google.golang.org/grpc@v1.51.0/clientconn.go:154 +0x1913
vitess.io/vitess/go/vt/grpcclient.DialContext({0x18bc7f8, 0xc000046038}, {0xc001878bd0, 0x11}, 0x0, {0xc007baa000, 0x1d7c, 0x52f485?})
	/Users/derek/go/pkg/mod/vitess.io/vitess@v0.15.1/go/vt/grpcclient/client.go:149 +0x78f
vitess.io/vitess/go/vt/grpcclient.Dial(...)
	/Users/derek/go/pkg/mod/vitess.io/vitess@v0.15.1/go/vt/grpcclient/client.go:98
vitess.io/vitess/go/vt/vtgate/grpcvtgateconn.DialWithOpts.func1({0x10dee60?, 0xc0005dc1b0?}, {0xc001878bd0, 0x11})
	/Users/derek/go/pkg/mod/vitess.io/vitess@v0.15.1/go/vt/vtgate/grpcvtgateconn/conn.go:89 +0x158
vitess.io/vitess/go/vt/vtgate/vtgateconn.DialProtocol({0x18bc7f8, 0xc000046038}, {0xc0063e6108, 0x4}, {0xc001878bd0, 0x11})
	/Users/derek/go/pkg/mod/vitess.io/vitess@v0.15.1/go/vt/vtgate/vtgateconn/vtgateconn.go:219 +0xec
vitess.io/vitess/go/vt/vitessdriver.(*conn).dial(0xc0042c8fa0)
	/Users/derek/go/pkg/mod/vitess.io/vitess@v0.15.1/go/vt/vitessdriver/driver.go:214 +0x3f
vitess.io/vitess/go/vt/vitessdriver.drv.Open({}, {0xc0000c0f80, 0x76})
	/Users/derek/go/pkg/mod/vitess.io/vitess@v0.15.1/go/vt/vitessdriver/driver.go:128 +0x185
```

--- a/go/vt/vtgate/grpcvtgateconn/conn.go
+++ b/go/vt/vtgate/grpcvtgateconn/conn.go
@@ -87,9 +87,14 @@ func Dial(opts ...grpc.DialOption) vtgateconn.DialerFunc {
 			return nil, err
 		}
 
-		opts = append(opts, opt)
+		// This dialer is shared and invoked concurrently for every new
+		// connection, so the captured opts must not be mutated: appending to
+		// it grows the slice on every dial and races between goroutines.
+		allOpts := make([]grpc.DialOption, 0, len(opts)+1)
+		allOpts = append(allOpts, opts...)
+		allOpts = append(allOpts, opt)
 
-		cc, err := grpcclient.DialContext(ctx, address, grpcclient.FailFast(failFast), opts...)
+		cc, err := grpcclient.DialContext(ctx, address, grpcclient.FailFast(failFast), allOpts...)
 		if err != nil {
 			return nil, err
 		}

--- a/go/vt/vtgate/grpcvtgateconn/conn_test.go
+++ b/go/vt/vtgate/grpcvtgateconn/conn_test.go
@@ -1,0 +1,79 @@
+/*
+Copyright 2026 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpcvtgateconn
+
+import (
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+)
+
+// TestDialDoesNotAccumulateOpts ensures the dialer returned by Dial does not
+// append to the caller-provided options slice. The dialer is registered once
+// and invoked for every new connection, so appending to the captured slice
+// grows it by one option per dial and mutates the caller's backing array.
+// See https://github.com/vitessio/vitess/issues/12068.
+func TestDialDoesNotAccumulateOpts(t *testing.T) {
+	// Extra capacity so a buggy append to the captured slice writes into
+	// the caller's backing array instead of reallocating.
+	callerOpts := make([]grpc.DialOption, 1, 8)
+	callerOpts[0] = grpc.WithUserAgent("grpcvtgateconn-test")
+
+	dialer := Dial(callerOpts...)
+	for range 3 {
+		conn, err := dialer(t.Context(), "localhost:0")
+		require.NoError(t, err)
+		conn.Close()
+	}
+
+	for i, opt := range callerOpts[len(callerOpts):cap(callerOpts)] {
+		assert.Nil(t, opt, "dialer mutated the caller's options slice at index %d", len(callerOpts)+i)
+	}
+}
+
+// TestDialConcurrent ensures concurrent invocations of a shared dialer do not
+// race on the captured options slice. database/sql dials new pool connections
+// from multiple goroutines; a race here tears dial option interface values
+// and panics inside grpc when they are applied.
+// See https://github.com/vitessio/vitess/issues/12067.
+func TestDialConcurrent(t *testing.T) {
+	dialer := Dial(grpc.WithUserAgent("grpcvtgateconn-test"))
+
+	// Each goroutine writes its own index, so no synchronization is needed;
+	// require lives in the test goroutine after Wait, where FailNow is valid.
+	const n = 10
+	errs := make([]error, n)
+	var wg sync.WaitGroup
+	for i := range n {
+		wg.Go(func() {
+			conn, err := dialer(t.Context(), "localhost:0")
+			if err != nil {
+				errs[i] = err
+				return
+			}
+			conn.Close()
+		})
+	}
+	wg.Wait()
+
+	for _, err := range errs {
+		require.NoError(t, err)
+	}
+}

